### PR TITLE
Improve experience with .flux.yaml files

### PR DESCRIPTION
**Pull request by squaremo**
- include more information like filenames in errors
 - correct the implementation of resourceID so it uses the name of the
   resource

These two address a comment https://github.com/fluxcd/flux/issues/2428#issuecomment-540031938 that fluxd should at least provide information about which file/resource is failing, when syncing goes wrong.

 - return an error if asked to update a manifest, but there are no (relevant) update commands in the config file

This addresses #2324 by making it more obvious what has failed (in that issue, the complaint can be interpreted as "nothing happened and I don't know why").

### CODE REVIEW

**Comment by squaremo** [comment]
@2opremio I've moved this out of draft, and done more work to surface mistakes with config files; PTAnotherL when you have a sec. :apple:

File: go.mod [hunk 1]
```diff
@@ -22,6 +22,7 @@ require (
 	github.com/pkg/errors v0.8.1
 	github.com/pkg/term v0.0.0-20190109203006-aa71e9d9e942
 	github.com/prometheus/client_golang v1.1.0
+	github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
```
Comment (2opremio) [inline go.mod]: Where did this come from? (just wondering)

File: pkg/manifests/configfile.go [hunk 2]
```diff
@@ -21,42 +23,60 @@ const (
...
-	Path       string
-	WorkingDir string
-	Version    int
+	Version int
 	// Only one of the following should be set simultaneously
 	CommandUpdated *CommandUpdated `yaml:"commandUpdated"`
 	PatchUpdated   *PatchUpdated   `yaml:"patchUpdated"`
+
+	path       string // the absolute path to the .flux.yaml
+	workingDir string // the absolute path to the dir in which to run commands or find a patch file
```
Comment (2opremio) [inline pkg/manifests/configfile.go]: Thanks! I should had thought about that!

File: pkg/manifests/configfile.go [hunk 3]
```diff
@@ -66,20 +86,119 @@ func NewConfigFile(path, workingDir string) (*ConfigFile, error) {
...
+				updaters[len(result)-1].Policy.Command,
+				result[len(result)-1].Error,
+				r.Source(),
+				result[len(result)-1].Output,
+			)
+			return false, err
+		}
+	}
+	// We assume that the update changed the resource. Alternatively, we could generate the resources
+	// again and compare the output, but that's expensive.
```
Comment (2opremio) [inline pkg/manifests/configfile.go]: Uhmm, we could simply use Git for this, I think. I wonder why I didn't think about it before (I know, not part of the PR)

File: pkg/manifests/configaware_test.go [hunk 4]
```diff
@@ -235,3 +235,44 @@ spec:
...
+           spec:
+             containers:
+             - name: greeter
+               image: quay.io/weaveworks/helloworld:master-a000001
+       ---
+       apiVersion: v1
+       kind: Namespace
+       metadata:
+         name: demo"
+  patchFile: patchfile.yaml
```
Comment (2opremio) [inline pkg/manifests/configaware_test.go]: Ideally, I think we should just reject this file and parse strictly (although returning an error when no updates are done is better and should have is certainly an improvement).

**Comment by 2opremio** [comment]
Very minor stuff, otherwise LGTM! 💯 (+ an extra 🥇 for the dispatching refactoring!)

File: go.mod (hunk 1 above)
Comment (squaremo) [reply go.mod]: I don't know, it keeps popping up whenever I run `make`; so whichever version of go I have, it thinks it's a dependency.

File: pkg/manifests/configfile.go (hunk 3 above)
Comment (squaremo) [reply pkg/manifests/configfile.go]: Right; I think this was probably ported over from the more circumspect implementation.

File: pkg/manifests/configaware_test.go (hunk 4 above)
Comment (squaremo) [reply pkg/manifests/configaware_test.go]: Yeah; validating against a schema was my next stop. While the particular error here would change, there would probably still be ways to have a config that doesn't quite work.
